extensions/binding/org.eclipse.smarthome.binding.digitalstrom/src/main/java/org/eclipse/smarthome/binding/digitalstrom/internal/lib/structure/devices/impl/DeviceImpl.java

```diff
@@ -0,0 +1,1859 @@
+        if (sensorEchoBox == null) {
+            sensorEchoBox = Collections.synchronizedList(new LinkedList<SensorEnum>());
+        }
+    }
+
+    @Override
+    public void disableSensorEchoBox() {
+        sensorEchoBox = null;
+    }
+
+    @Override
+    public boolean isSensorEchoBoxEnabled() {
+        return sensorEchoBox != null;
+    }
+
+    @Override
+    public DeviceSensorValue getDeviceSensorValue(SensorEnum sensorType) {
+        if (sensorType != null) {
+            for (DeviceSensorValue devSenVal : deviceSensorValues) {
+                if (devSenVal.getSensorType().equals(sensorType)) {
+                    return devSenVal;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public DeviceSensorValue getDeviceSensorValue(Short sensorIndex) {
+        if (sensorIndex != null) {
+            for (DeviceSensorValue devSenVal : deviceSensorValues) {
+                if (devSenVal.getSensorIndex().equals(sensorIndex)) {
+                    return devSenVal;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Short getSensorIndex(SensorEnum sensorType) {
+        if (sensorType != null) {
+            DeviceSensorValue devSenVal = getDeviceSensorValue(sensorType);
+            return devSenVal != null /* && devSenVal.getValid() */ ? devSenVal.getSensorIndex() : null;
+        }
+        return null;
+    }
+
+    @Override
+    public SensorEnum getSensorType(Short sensorIndex) {
+        if (sensorIndex != null) {
+            DeviceSensorValue devSenVal = getDeviceSensorValue(sensorIndex);
+            return devSenVal != null /* && devSenVal.getValid() */ ? devSenVal.getSensorType() : null;
+        }
+        return null;
+    }
+
+    @Override
+    public Integer getDsSensorValue(SensorEnum sensorType) {
+        return getDsSensorValue((Object) sensorType);
+    }
+
+    @Override
+    public Integer getDsSensorValue(Short sensorIndex) {
+        return getDsSensorValue((Object) sensorIndex);
+    }
+
+    @Override
+    public Float getFloatSensorValue(Short sensorIndex) {
+        return getFloatSensorValue((Object) sensorIndex);
+    }
+
+    @Override
+    public Float getFloatSensorValue(SensorEnum sensorType) {
+        return getFloatSensorValue((Object) sensorType);
+    }
+
+    @Override
+    public boolean setFloatSensorValue(SensorEnum sensorType, Float floatSensorValue) {
+        return checkAndSetSensorValue(sensorType, null, floatSensorValue);
+    }
+
+    @Override
+    public boolean setFloatSensorValue(Short sensorIndex, Float floatSensorValue) {
+        return checkAndSetSensorValue(sensorIndex, null, floatSensorValue);
+    }
+
+    @Override
+    public boolean setDsSensorValue(Short sensorIndex, Integer dSSensorValue) {
+        return checkAndSetSensorValue(sensorIndex, dSSensorValue, null);
+    }
+
+    @Override
+    public boolean setDsSensorValue(SensorEnum sensorType, Integer dSSensorValue) {
```

Comment:
Why are here 2 functions, one for type and one for index? I guess type and index are related in a sense that both should always be changed together and kept in sync? If that is the case I would clearly vote for creating only ONE method to set both... Same applies to the other setters here.

Reply:
No, some functions or events from digitalSTROM use or return the index and some other the type.

And the maping of the type to the index can be different by every device-type.